### PR TITLE
std: allow overriding install path of artifacts

### PR DESCRIPTION
**Pull request by ifreund**
This is necessary when, for example, writing a PAM module which should
be installed to lib/security/module_name.so.

Currently the zig build system forces shared libraries to be prefixed
with "lib" and versions the so files as well as creating symlinks.
Furthermore, artifacts can currently only be installed to the bin or
lib directories, not sub-directories thereof.

Allowing the user to manually specify the install path allows maximum
flexibility, solving this use case and more.

### CODE REVIEW

**Comment by ifreund** [comment]
This feature would allow me to merge this [commit](https://github.com/ifreund/rundird/commit/8efa886f355bda1000f89cccf56178d4c23e69f0) replacing my current Makefile with a `build.zig`. However, the current implementation may be considered a little heavy-handed, suggestions on how to improve it are of course welcome.

**Comment by andrewrk** [comment]
I definitely want to solve this use case and unblock you - thank you for bringing this up. I haven't looked very deeply into this, but what are your thoughts on extending this to support the security subdir?

https://github.com/ziglang/zig/blob/800c5de2ae527cdefaa5567720dab2f899244897/lib/std/build.zig#L2618-L2623

**Comment by ifreund** [comment]
>I definitely want to solve this use case and unblock you - thank you for bringing this up. I haven't looked very deeply into this, but what are your thoughts on a different approach of extending InstallDir to support the security subdir?

Hmm, I don't think that we should have to teach the zig build system about every possible sub directory that one might want to install to, instead there should be a way to take control and install artifacts to any arbitrary path.

That also wouldn't solve the issue with the naming of the so file, but I see now that #2231 and #2230 would.

As a way to move forward with this I think I'll give implementing those a go and then revisit this.

**Comment by andrewrk** [comment]
> Hmm, I don't think that we should have to teach the zig build system about every possible sub directory that one might want to install to, instead there should be a way to take control and install artifacts to any arbitrary path.

That makes sense - what about making `InstallDir` a tagged union and adding a "custom" tag?

File: lib/std/build.zig [hunk 1]
```diff
@@ -1212,6 +1213,8 @@ pub const LibExeObjStep = struct {
     is_linking_libc: bool = false,
     vcpkg_bin_path: ?[]const u8 = null,
 
+    /// This may be set in order to override the default install directory
+    override_dest_dir: ?InstallDir,
```
Comment (ifreund) [inline lib/std/build.zig:1217]: Doesn't really seem like a function to set this variable is necessary, users of the api can just set the field directly.

File: lib/std/build.zig [hunk 2]
```diff
@@ -2309,17 +2313,17 @@ pub const InstallArtifactStep = struct {
             .builder = builder,
             .step = Step.init(.InstallArtifact, builder.fmt("install {}", .{artifact.step.name}), builder.allocator, make),
             .artifact = artifact,
-            .dest_dir = switch (artifact.kind) {
+            .dest_dir = artifact.override_dest_dir orelse switch (artifact.kind) {
                 .Obj => unreachable,
                 .Test => unreachable,
-                .Exe => .Bin,
-                .Lib => .Lib,
+                .Exe => InstallDir{ .Bin = {} },
```
Comment (andrewrk) [inline lib/std/build.zig:2319]: I think you're working around a stage1 issue here, but just to let you know it's planned for `.Bin` to coerce to the tagged union here, which would make these kind of code reworkings nicer in the future.

**Comment by ifreund** [comment]
> Looks good - and you made sure this solves your use case in particular, right?

It will once #2230 and #2231 are solved, with just this PR my build.zig installs to `lib/security/libpam_rundird.so.0.0.0` (plus symlinks) while what I need in the end is `lib/security/pam_rundird.so`.

#6315 makes further progress towards this.

File: lib/std/build.zig (hunk 2 above)
Comment (ifreund) [reply lib/std/build.zig:2319]: Yeah, that seems to be the case unfortunately, I played with it a bit but this was the cleanest it got.

**Pull request merged**